### ISSUE TITLE
strange interaction of Shift with lines&curves

### ISSUE THREAD

**Issue by jeisner** [mypaint/mypaint#456]
Regarding the lines&curves tool and the connected lines tool:

The documentation (in the hint line at the bottom of the window) says "Shift adds curves, Ctrl constrains angle."  However, Shift seems to constrain the angle too, while one is positioning the line.  I assume this is deliberate (at least for now, until another good use for Shift arises)?

More strangely, if you temporarily switch to the lines & curves mode by holding Shift while in freehand mode, then 
* *Releasing* Shift (while positioning the line) will now constrain the angle.  This seems weird enough to be unintentional.  
* Although the documentation still says "Shift adds curves, Ctrl constrains angle," there doesn't actually seem to be a way to add curves in this case.  Or have I missed it?

**Comment by achadwick** [comment]
This sounds like something it would be better to cover in the documentation wiki pages.

We can't change the strings within the program now before the 1.2.0 release, and I'm swamped  fighting fires elsewhere (mostly new upstream issues from GTK :cold_sweat: )

**Comment by jeisner** [comment]
Ok.  But note that two items in this issue could be implementation bugs, rather than a documentation bugs:

If you press Shift in freehand mode, start a stroke, and then release Shift,  the stroke angle is now constrained for some reason.  This is weird.

If you're in lines&curves mode or connected lines mode,  then Shift constrains angles.  This behavior is undocumented, but may be intentional -- not sure.

**Comment by achadwick** [comment]
> If you press Shift in freehand mode, start a stroke, and then release Shift, the stroke angle is now constrained for some reason. This is weird.

Yeah, the logic is well weird there. Shift + paint starts straight line mode, and *altering the state of shift or ctrl* toggles angle snapping.

This is weird, so it should be documented :)

> If you're in lines&curves mode or connected lines mode, then Shift constrains angles. This behavior is undocumented, but may be intentional -- not sure.

This is intentional. Really the docs just need to give the story in an understandable way :/

**Comment by jeisner** [comment]
On Thu, Oct 22, 2015 at 12:39 PM, Andrew Chadwick <notifications@github.com>
wrote:

> If you press Shift in freehand mode, start a stroke, and then release
> Shift, the stroke angle is now constrained for some reason. This is weird.
>
> Yeah, the logic is well weird there. Shift + paint starts straight line
> mode, and *altering the state of shift or ctrl* toggles angle snapping.
>
> This is weird, so it should be documented :)
>
My humble opinion is that it should actually be changed. Releasing shift in
this setting should have no effect.  But pressing shift again should then
turn on angle snapping.

In other words, the behavior should be that angle snapping is on when shift
or ctrl is depressed, unless it's only depressed because that was how the
mode was entered in the first place.

The current behavior is confusing, and also means that the user has to keep
shift depressed for the entire stroke, rather than just as part of the
gesture that starts the stroke.  It's nice to be able to let go of the
shift key without the (unexpected) side effect of angle snapping.

**Comment by jeisner** [comment]
p.s.  The description you gave of the "weird logic" may not be quite accurate.  Here's some more confusing interactions:

Freehand mode.
Hold down shift.
Start a stroke.
Release shift.  (Now you're in angle snapping mode.)

Problem 1: Your description suggests that angle snapping can now be turned off by pressing either Shift or Ctrl.  However, I find that this is only true for Shift.  Pressing and releasing Ctrl has *no* effect on angle snapping mode.

Problem 2 (much worse): When you finish the line, you're now still in lines&curves mode, as if you'd entered it by pressing K!   Even though you are no longer holding shift.  So if you try to continue with a freehand stroke, you'll discover it's surprisingly a line.  And if you try to continue with a line by pressing shift again, you'll discover that it surprisingly turns your previous line into a curve.   I have *often* run afoul of this behavior in earlier versions.

**Comment by jeisner** [comment]
See some related discussion on #458.

**Comment by zb13y** [comment]
> Problem 2 (much worse): When you finish the line, you're now still in lines&curves mode, as if you'd entered it by pressing K! Even though you are no longer holding shift. So if you try to continue with a freehand stroke, you'll discover it's surprisingly a line. And if you try to continue with a line by pressing shift again, you'll discover that it surprisingly turns your previous line into a curve. I have often run afoul of this behavior in earlier versions.
This must be changed, not documented. Maybe something like this: when user holding Shift - user starts to draw straight line, when user releases Shift - MyPaint reverts to Freehand Mode. And for curved lines user changes mode to Line and Curves. So holding Shift used only for straight lines?
